### ISSUE TITLE
Make the command-line libraries optional

### ISSUE THREAD

**Issue by Kixiron** [vmware/differential-datalog#740]
Currently the `cmd_parser` generated crate is included in every build of a generated project when they aren't needed for users who use the generated code solely as a library. The `cmd_parser` crate and the cli interface add a number of unnecessary dependencies for this kind of use case:

- `rustop`
- `nom`
- `num`
- `rustyline`

Similarly the `ovsdb`-related libraries and generated crates, the generated `main.rs` and the generated C headers can all be toggled for users who just don't need them. `ovsdb` itself adds `serde_json` and `uuid`.
All in all, this can significantly reduce compile times and the dependency load of users who just want a library.

There's also more dependency culling that can be done, `serde` and `abomination` can be an optional features (like `flatbuf`) which removes the dependencies *and* the generated code which is a double-win, `twox-hash` doesn't seem to be used anywhere so it can probably be removed, `num-traits` could possibly be removed

**Comment by ryzhyk** [comment]
Some of these are definitely doable.  In particular, ovsdb should be feature-gated. `cmd-parser` can also be generated on-demand, but you typically don't get very far without it.  Even if your program does not use it directy, it is very helpful in replay debugging DDlog code.

`abomonation` is required by differential dataflow, so that cannot be opted out of.
`twox-hash`, `num`, and `num-traits` are, I believe, used by the DDlog runtime to implement its built-in types, and so are also required.

`serde` can probably be made optional, although a lot of generated and library code relies on it, so we'll have to do some careful surgery to allow this, but the pay-off will likely to be most significant for programs that don't need it (see below).

Having said that, most of these dependencies will not reduce compilation time significantly.  The reason DDlog programs take so long to compile is that LTO takes forever for DDlog-generated data types, which is in turn because each of them auto-derives a bunch of traits. This is a known problem in Rust. In the past we put a lot of effort in optimizing compilations times, but I think we need to do another round of that, as this is also affecting some of our internal projects significantly.

PS. Apologies about being slow to actually resolve any of these issues. We're kind of overwhelmed right now.  I expect that the Rust API is the one that will become a blocker for you soon (if not already).  I'll try to find time to help with that.  Or did you already figure it out yourself?

**Comment by Kixiron** [comment]
Yes, I actually looked into `twox-hash` more and it's used in the std's hashing functions, but `fnv` is already included as a dependency so that can likely be used
The cli stuff is definitely useful, but there's also situations where it's not at all used, like in a library's usage. It's good for debug mode, but it's not needed for release
The api is a little rocky, but I've definitely used worse. I've got a few ideas for improving it, but I need to get more accustomed to using everything first

Y'all are totally fine on response times, I've been feeling guilty that I'm just opening issues and not helping any, but ghc keeps segfaulting when I try and build the project so my workflow is *interesting* to say the least

**Comment by ryzhyk** [comment]
We wanted cryprographic hash for that function, which `fnv` isn't. Having said that, we should probably have a separate library for crypto-strength hashing. 

> Y'all are totally fine on response times, I've been feeling guilty that I'm just opening issues and not helping any, but ghc keeps segfaulting when I try and build the project so my workflow is interesting to say the least

Hmm. Can you try changing `resolver` to `lts-16.9` in `stack.yml` and re-run `stack install`?

**Comment by Kixiron** [comment]
Still no dice
```
❯ stack install
WARNING: Ignoring regex-posix's bounds on base (<0 && >=4.3 && <4.15); using base-4.13.0.0.
Reason: trusting snapshot over cabal file dependency information.
[1 of 2] Compiling Main             ( C:\\sr\setup-exe-src\setup-Z6RU0evB.hs, C:\\sr\setup-exe-src\setup-Z6RU0evB.o )

Access violation in generated code when writing 0x0

 Attempting to reconstruct a stack trace...

   Frame        Code address
 * 0x461da80    0x37f7b66 C:\Users\Chase\AppData\Local\Programs\stack\x86_64-windows\ghc-8.8.3\bin\ghc.exe+0x33f7b66
 * 0x461da88    0x3277bb9 C:\Users\Chase\AppData\Local\Programs\stack\x86_64-windows\ghc-8.8.3\bin\ghc.exe+0x2e77bb9
 * 0x461da90    0x3
 * 0x461da98    0x6b95fa1
 * 0x461daa0    0x3d80c90 C:\Users\Chase\AppData\Local\Programs\stack\x86_64-windows\ghc-8.8.3\bin\ghc.exe+0x3980c90
 * 0x461daa8    0x6a35b60


--  While building simple Setup.hs using:
      C:\Users\Chase\AppData\Local\Programs\stack\x86_64-windows\ghc-8.8.3\bin\ghc-8.8.3.EXE -rtsopts -threaded -clear-package-db -global-package-db -hide-all-packages -package base -main-is StackSetupShim.mainOverride -package Cabal-3.0.1.0 C:\sr\setup-exe-src\setup-Z6RU0evB.hs C:\sr\setup-exe-src\setup-shim-Z6RU0evB.hs -o C:\sr\setup-exe-cache\x86_64-windows\tmp-Cabal-simple_Z6RU0evB_3.0.1.0_ghc-8.8.3.exe
    Process exited with code: ExitFailure 11
```

**Comment by ryzhyk** [comment]
Looks like an instance of this: https://gitlab.haskell.org/ghc/ghc/-/issues/17926

**Comment by ryzhyk** [comment]
Can you try this in `stack yaml`?

```
resolver: lts-16.9
compiler: ghc-8.8.4
```

**Comment by Kixiron** [comment]
After much trial and error, I got `ghc-8.10.2` to work, but now I get actual [build errors](https://paste.rs/8OB)

**Comment by ryzhyk** [comment]
It looks like auto-derived `Serialize`/`Deserialize` traits are responsible to much of the compilation time.

In a large project with hundreds of user-defined types compilation time goes down from 8m38s to 4m47s when I manually remove all the serialization stuff.  Of the remaining 5 minutes, more than half is probably due to crates that never change and so do not affect re-compilation times.

So yes, making serialization optional is a good idea for projects that don't need it; although this is not a small change, as various  parts of DDlog assume that all types implement these traits.  More importantly, there are projects that actually use serialization that I would still like to be able to compile faster.

One approach we could try is to get DDlog to generate the implementations of `Serialize`/`Deserialize` for user-defined types instead of `#derive`-ing them. This way we will be able to annotate `serialize`, `deserialize` methods with `#inline(never)` to, hopefully, speed up the LTO pass (which is where rustc spends most of the time when compiling ddlog-generated code).

**Comment by Kixiron** [comment]
That sounds right, especially considering my experience with derive macros (especially ones as complex as serde's). Gaming lto is also a good idea, maybe not with `#[inline(never)]` as it's overly aggressive and can seriously impact perf, but maybe with no annotations. Once the linker sees things not being used/the frequency that they are used it'll likely outline them anyways.

**Comment by ryzhyk** [comment]
Can you clarify what you mean by "no annotations"?

**Comment by Kixiron** [comment]
Just the absence of an attribute, instead of `#[inline]` or `#[inline(never)]` simply put nothing

**Comment by ryzhyk** [comment]
Understood.  I think this is the default serde behavior, so it won't improve things for us. The plan was to actually sacrifice a bunch of serialization/deserialization performance for compilation speed.  Since this is auto-generated code, we can always have a compiler switch to remove `#[inline(never)]` in production builds.

Having said that, I am not 100% sure this will do the trick.  Plus generating our own `Serialize`/`Deserialize` implementation is not a lot of fun and will probably require ripping out a bunch of code from `serde`.  I am currently trying to understand exactly what effects compilation time and size of serde-generated code.

**Comment by Kixiron** [comment]
[Rust's self profiler](https://blog.rust-lang.org/inside-rust/2020/02/25/intro-rustc-self-profile.html) can probably give good insight

**Comment by mbudiu-vmw** [comment]
It's not obvious to me that if you generate the code yourself it will be faster to compile.

**Comment by ryzhyk** [comment]
@mbudiu-vmw, yes. I initially thought that disabling inlining will help, but looking more closely into the generated code, it's more complex than that.

**Comment by Kixiron** [comment]
Procedural macros require what's essentially an extra build; Proc-macros themselves are compiled separately as an executable (of sorts), along with all of their dependencies. This new executable is then run at compile time and the resulting code injected into your crate
As of now, procs don't have very good incrementality, which means they run much more frequently than they should. In addition to that, the generated code is often pretty gnarly since it's all generated, and in mass quantities that means there's just simply a lot of mass to churn through, both for the end compiler and the procedural macro that's doing the code generation

**Comment by ryzhyk** [comment]
Closed via #749
